tools/config/__init__.py

```diff
@@ -600,15 +600,26 @@ def _get_cmsis_part(self):
                                   "arm_pack_manager index.")
         return cache.index[self.target.device_name]
 
-    def _get_mem_specs(self, memories, cmsis_part, exception_text):
-        for memory in memories:
-            try:
-                size = cmsis_part['memory'][memory]['size']
-                start = cmsis_part['memory'][memory]['start']
-                return (start, size)
-            except KeyError:
-                continue
-        raise ConfigException(exception_text)
+    @staticmethod
+    def _memory_ordering(memory):
+        return (memory['default'], memory['size'], memory['start'])
+
+    def _get_mem_specs(self, permissions, cmsis_part):
+        all_matching_memories = [
+            memory for memory in cmsis_part['memories'].values()
+            if all(memory['access'][perm] for perm in permissions)
+        ]
+        if all_matching_memories:
+            memory = sorted(
+                all_matching_memories, key=self._memory_ordering
```

Comment:
I'm a bit tripped up on the `key` function here. Is this essentially doing the following for the comparison:

"
For each `memory` in `all_matching_memories`, the one that is the "least" is the one that:

1. Has the smallest value for `memory['default']`
2. If they are the same, pick the smallest value for `memory['size']`
3. If that is the same, pick the smallest value for `memory['start']`
"

Reply:
@bridadan I was under the impression that it would be largest first. If I need a quick reverse to make them all "largest" then that's what I'll do.

Reply:
`sorted` will return an ascending order sorted list by default: https://docs.python.org/3/howto/sorting.html. So that would mean the first one in the list would be the "smallest" if I'm interpreting the doc correctly.

Reply:
Right. Adding a reverse now.